### PR TITLE
Don't skip all-success on deps failure.

### LINKED ISSUE #11569: Enforce protected branches
GitHub has an issue in which when a dependent job fails, the job depending on it will skip.

We want the downstream job to also fail.

In this way, in GitHub branch protection API we can define a single job for a workflow file.

In this way,  each time you modify a job in a worflow file, you don't have to also update the GitHub UI config.

### PR DESCRIPTION
**Pull request by adiroiban**
## Scope and purpose

Fixes #11569 

This applies the changes created by @altendky  from here https://github.com/Chia-Network/chia-blockchain/pull/11880/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R79 

The code was already tested in towncrier here https://github.com/twisted/towncrier/pull/392

Note that this only handles the jobs defined inside our `test.yaml` files

codecov.io, Azure Pipelines, Read The Docs, and pre-commit are external services and are not covered by this change.

In theory, we can enabled probot-settings app and have all the repo settings as YAML 
https://github.com/repository-settings/app

Anyone with push access to the repo will be admin and could change the branch protection via YAML config.

```
Author: adiroiban
Reviewer: 
Fixes #11569

Fail and don't skip CI branch protection when a job fails.
```

### CODE REVIEW

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -257,12 +257,29 @@ jobs:
...
+      - name: Require all successes
+        shell: python
+        env:
+          RESULTS: ${{ toJSON(needs.*.result) }}
+        run: |
+          import json
+          import os
+          import sys
+          results = json.loads(os.environ["RESULTS"])
+          sys.exit(0 if all(result == "success" for result in results) else 1)
```
Comment (altendky) [inline .github/workflows/test.yaml:285]: As I see this show up in more places, maybe it should be made into an action?  Or the whole job into a reusable workflow.  In this case the reusable workflows are used to get around the 256 job per matrix limit, but it is still an example with the `needs:` in use.

https://github.com/Chia-Network/chia-blockchain/blob/9a4316f96eeb92a06d5ceaf94387d349c5ea9537/.github/workflows/test.yml#L45-L55
```yaml
  macos:
    uses: ./.github/workflows/test-single.yml
    needs: configure
    with:
      emoji: 🍎
      matrix: macos
      name: macOS
      file_name: macos
      concurrency_name: macos
      configuration: ${{ needs.configure.outputs.configuration }}
      runs-on: macos-latest
```

**Review by altendky — APPROVED** [review]
It reads right at least.  The reusability comment can be taken on separately if we get to that.

File: .github/workflows/test.yaml (hunk 1 above)
Comment (adiroiban) [reply .github/workflows/test.yaml:285]: I have merged this as it should be better than what we have now.

I think that a shared action should be good enough.

--------

I am not sure how workflow interdependencies work.

**Comment by adiroiban** [comment]
Many thanks for the review! :)

File: .github/workflows/test.yaml (hunk 1 above)
Comment (altendky) [reply .github/workflows/test.yaml:285]: I did make a first try at shared workflow.  You do have to pass the `needs:` and code the `if: always()` in the repeated location so...  It's something, but not what I had hoped.  https://github.com/Chia-Network/chia-blockchain/pull/11880

Comment (adiroiban) [reply .github/workflows/test.yaml:285]: Thanks... I think that is nice, but for Twisted is a bit complicated.

I mean the current code, should be easier to read and understand by average Twisted dev that is not an expert into GiHub Actions.

Comment (webknjaz) [reply .github/workflows/test.yaml:285]: > As I see this show up in more places, maybe it should be made into an action?

Yes: https://github.com/re-actors/alls-green

**Pull request merged**